### ISSUE TITLE
Panic during training, burn-ir/handle.rs

### ISSUE THREAD

**Issue by eblack-leaf** [tracel-ai/burn#4582]
---                                                                                                                                                                                                                        
  ## Bug: `no_grad()` embedding parameters crash training at epoch boundary on wgpu/fusion backend                                                                                                                           
                                                                                                                                                                                                                             
  **burn version:** 0.20.1                                                                                                                                                                                                   
  **backend:** `Autodiff<Fusion<WgpuRuntime>>`                                                                                                                                                                               
  **crate:** `burn-ir`, `burn-train`                                                                                                                                                                                         
                                                                                                                                                                                                                             
  ---                                                                                                                                                                                                                        
                                                                                                                                                                                                                             
  ### Description                                                                                                                                                                                                            
                                                                                                                                                                                                                             
  When an `Embedding` layer is frozen with `.no_grad()` and the model is trained with
  `SupervisedTraining` + `with_file_checkpointer`, training panics at the start of the
  **third epoch** (iteration 1–2) with:

  panicked at burn-ir-0.20.1/src/handle.rs:88:
  Should have handle for tensor TensorId { value: ... }

  This causes the async checkpointer thread to receive a `SendError` on its channel,
  which then panics again in its destructor, producing a double-panic abort:

  panicked at async_checkpoint.rs:123: Can send message to checkpointer thread.
  panicked at async_checkpoint.rs:162: Can send the end message to the checkpointer thread.
  panicked at panicking.rs:233: panic in a destructor during cleanup
  Aborted (core dumped)

  The failure is **100% reproducible** and always triggers at the same point: the first
  training iteration after epoch 2's validation + checkpoint.

  ---

  ### Root cause

  `get_handle` in `burn-ir` has two code paths for the `TensorStatus`:

  ```rust
  TensorStatus::ReadOnly  => { self.handles.insert(id, Handle::Existing(handle.clone())); handle }
  TensorStatus::ReadWrite => handle  // ← handle is consumed; NOT re-inserted

  When SupervisedTraining saves a checkpoint after epoch 2 validation, the async
  checkpointer serialises the model's parameters to disk. For a no_grad() embedding,
  this involves a ReadWrite drain of the weight tensor's GPU buffer — which
  removes the handle from the map and does not re-insert it.

  Because the embedding has require_grad = false, the fusion backend considers the
  tensor "done" after this drain: it is not referenced in any pending autodiff graph
  and will not be needed for a backward pass. The handle is therefore not
  re-registered.

  When epoch 3 starts and embedding.forward() is dispatched to the fusion backend,
  it looks up the embedding weight's TensorId — which is no longer in the handle
  map — and panics.

  The sequence in the experiment log confirms this precisely:

  Executing validation step for epoch 2   ← epoch 2 validation runs fine
  Executing training step for epoch 3
  Iteration 1
  ERROR: panicked … Should have handle for tensor TensorId { value: … }

  ---
  Minimal reproduction

  // model with frozen GloVe embedding
  let embedding = EmbeddingConfig::new(vocab_size, embed_dim)
      .init(device)
      .no_grad();               // ← triggers the bug

  // train with file checkpointer on wgpu backend
  SupervisedTraining::new(dir, train_loader, val_loader)
      .with_file_checkpointer(CompactRecorder::new())
      .num_epochs(10)
      .launch(learner);
  // → panics at start of epoch 3

  ---
  Why no_grad() triggers this but regular params don't

  For a parameter with require_grad = true, the fusion backend keeps the handle
  alive because it expects the tensor to participate in a future backward pass (the
  handle will be needed for the gradient computation). The checkpoint read is done
  with ReadOnly status, which clones and re-inserts the handle.

  For require_grad = false, the fusion backend treats the tensor as a constant with
  no future autodiff use. The checkpoint read uses ReadWrite (draining the buffer
  to CPU), consuming the handle permanently.

  ---
  Workaround

  Do not use .no_grad() on parameters that must persist across checkpoint
  boundaries on the fusion backend. Instead, let gradients flow normally and
  strip the parameter's ParamId from GradientsParams before the optimizer step:

  fn step(&self, item: Input) -> TrainOutput<Output> {
      let out = self.forward_classification(item.tokens, item.labels);
      let mut grads = GradientsParams::from_grads(out.loss.backward(), self);
      if self.freeze_embeddings {
          for id in list_param_ids::<_, B>(&self.embedding) {
              let _ = grads.remove::<B::InnerBackend, 2>(id);
          }
      }
      TrainOutput { grads, item: out }
  }

  This keeps require_grad = true on the embedding weight, so the fusion backend
  maintains its handle across checkpoints. The optimizer never receives a gradient
  for the embedding, so the weights remain frozen.

  ---
  Expected behaviour

  no_grad() parameters should either:

  1. have their handles preserved across checkpoint serialisation on the fusion
  backend (using ReadOnly rather than ReadWrite during the drain), or
  2. be re-registered in the handle map after the checkpoint read completes.

  ---
  Environment

  burn        = "0.20.1"
  burn-train  = "0.20.1"
  burn-ir     = "0.20.1"
  backend     = wgpu (fusion enabled)
  OS          = Linux 6.17.0 x86_64

  ---

**Comment by antimora** [comment]
Same panic family as #4347. Several fixes landed since (#4962, #5282, #5400, #5535). Could you retest on current main (or 0.22 once it's out)? If we don't hear back in a few weeks we'll assume it's fixed and close this.
